router/dataplane_concurrency_model.gobra

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 ETH Zurich
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +gobra
+
+package router
+
+import (
+	"sync"
+	io "github.com/scionproto/scion/verification/io"
+)
+
+type SharedArg struct {
+	Place *io.Place // Existential for the current place
+	State *io.IO_dp3s_state_local // Existential for the current model state
+	IBufY, OBufY ElemRA // Parameters of the resoure algebra (is not necessary to understand to verify I/O spec)
+}
+
+pred SharedInv(ghost dp io.DataPlaneSpec, ghost y SharedArg) {
+	acc(y.Place) && acc(y.State) && // existentials are expressed using ghost pointers
+	io.token(*y.Place) && dp.dp3s_iospec_ordered(*y.State, *y.Place) && 
+	ElemAuth((*y.State).ibuf, y.IBufY) && ElemAuth((*y.State).obuf, y.OBufY)
+}
+
+// initialize the shared invariant:
+ghost 
+requires io.token(p) && dp.dp3s_iospec_ordered(s, p)
+ensures  m.LockP() && m.LockInv() == SharedInv!< dp, y !>;
+func InitSharedInv(ghost dp io.DataPlaneSpec, ghost p io.Place, ghost s io.IO_dp3s_state_local) (ghost m *sync.Mutex, ghost y SharedArg) {
+	mV@ := sync.Mutex{}
+	m = &mV
+	pE@ := p
+	sE@ := s
+	yI := InitElemAuth(s.ibuf)
+	yO := InitElemAuth(s.obuf)
+	y := SharedArg{&pE, &sE, yI, yO}
+	fold SharedInv!< dp, y !>()
+	m.SetInv(SharedInv!< dp, y !>)
+}
+
+/////////////////////////////////////// Lib ///////////////////////////////////////
+
+pred Prophecy(int)
+
+ghost
+ensures 0 <= expectedM && Prophecy(expectedM)
+decreases
+func AllocProphecy() (expectedM int)
+
+/////////////////////////////////////// MultiBio ///////////////////////////////////////
+
+pred MultiReadBio(ghost t io.Place, n int) {
+  (n > 0 ==> io.CBioIO_bio3s_recv(t) && MultiReadBio(io.dp3s_iospec_bio3s_recv_T(t), n-1))
+}
+
+ghost
+requires m <= n && MultiReadBio(t, n)
+decreases m
+pure func MutliReadBio_next(ghost t io.Place, n, m int) (ghost tn io.Place) {
+  return m <= 0 ? t : unfolding MultiReadBio(t, n) in 
+    MutliReadBio_next(io.dp3s_iospec_bio3s_recv_T(t), n-1, m-1)
+}
+
+ghost 
+requires m <= n && MultiReadBio(t, n)
+decreases m
+pure func MutliReadBio_correctIfs(ghost t io.Place, n, m int, ghost k Key) bool {
```

Comment:
same comment about the name of the function. Also, `n` and `m` are not very descriptive. What does this function do? Can we add a comment explaining that?